docs/designs/serving_design.md

```diff
@@ -0,0 +1,103 @@
+
+Large size model  
+A single serving process will run out of memory while loading the model. We partition the model variables into multiple shards, store them in distributed parameter server for serving. The inference engine will execute the serving graph, query the variable values from the distributed parameter server as needed and finish the calculation. It's necessary to upgrade the serving framework to support this. **We will discuss this case in a separate design in the next step.**
+
+In this article, we want to achieve these three goals in ElasticDL:
+
+1. Guarantee consistency between training and serving for input data preprocessing.
+2. Export the model with elasticdl.layers.Embedding to SavedModel.
+3. Execute task to save model with fault-tolerance.
+
+## Guarantee consistency between training and serving for input data preprocessing
+
+ElasticDL is a distributed deep learning framework based on TensorFlow 2.0 eager execution. In ElasticDL, we use tf.data.Dataset to create input pipeline for training. The recommended way to preprocess data from tf.data.Dataset is to use [feature columns](https://www.tensorflow.org/tutorials/structured_data/feature_columns) in TensorFlow. What's more, tf.saved_model.save will save the defined feature columns with the model. So tf-serving will use the same preprocessing logic with training to make inference.
+
+Define a Keras model with feature columns:
+
+```python
+def get_feature_columns():
+    age = tf.feature_column.numeric_column("age", dtype=tf.int64)
+    education = tf.feature_column.categorical_column_with_hash_bucket(
+        'education', hash_bucket_size=4)
+    education_one_hot = tf.feature_column.indicator_column(education)
+    return [age, education_one_hot]
+
+def get_input_layer():
+    input_layers = {}
+    input_layers['age'] = tf.keras.layers.Input(name='age', shape=(1,), dtype=tf.int64)
+    input_layers['education'] = tf.keras.layers.Input(name='education', shape=(1,), dtype=tf.string)
+    return input_layers
+
+def custom_model(feature_columns):
+    input_layers = get_input_layer()
+    dense_feature = tf.keras.layers.DenseFeatures(feature_columns=feat_cols)(input_layers)
+    dense = tf.keras.layers.Dense(10, activation='relu')(dense_feature)
+    dense = tf.keras.layers.Dense(1, activation='sigmoid')(dense)
+    return tf.keras.models.Model(inputs=input_layers, outputs=dense)
+```
+
+Although all feature columns in TensorFlow can be used in ElasticDL, tf.feature_column.embedding_column is not recommended in ElasticDL. Because the embedding_column has a variable containing a large embedding table. In eager execution the model must get all the embedding parameters to train. It will bring a large inter-process communication overhead.
```

Comment:
It looks like we miss some sentences at the end. We don't recommend users to use tf.feature_column.embedding_column, based on some reasons. So users should use ..? We can add one more sentence to connect with the next section. Or we just remove this paragraph, as I saw you put a detailed explanation at the beginning of the next section.

Reply:
When a categorical column has many possible values, tf.feature_column.embedding_column will be used to wrap categorical columns for tf.keras.layers.Densefeature.  But, the embedding layer cannot do it for categorical columns. Now, we don't find a solution to feed the keras model with the output of a categorical column when a categorical column has many possible values. Here, I just want to remind users of this issue.